lib/mercury/cps.rb

```diff
@@ -0,0 +1,97 @@
+require 'mercury/cps/seq'
+require 'mercury/cps/seq_with_let'
+require 'mercury/cps/methods'
+require 'mercury/utils'
+
+# Async IO often involves CPS (continuation-passing style)
+# code, where the continuation (a.k.a. "callback") is passed
+# to a function to be invoked at a later time. CPS style
+# can result in deep lexical nesting making code difficult
+# to read.
+# This monad hides the CPS plumbing, which allows code
+# to be written in a flat style with no visible continuation
+# passing. It basically wraps a CPS proc with methods for
+# composing them.
+# See http://codon.com/refactoring-ruby-with-monads
+class Cps
+  attr_reader :cps
+
+  # @param [Proc] cps a CPS proc (signature: *args, &k)
+  def initialize(&cps)
+    @cps = cps
+  end
+
+  # Applies the wrapped proc. If the CPS return value
+  # is not needed, the continuation k may be omitted.
+  # Returns the return value of the continuation.
+  def run(*args, &k)
+    k ||= proc{|x| x}
+    cps.call(*args, &k)
+  end
+
+  # The "bind" operation; composes two Cps
+  # @param [Proc] pm a proc that takes the output of this
+  #   Cps and returns a Cps
+  def and_then(&pm)
+    Cps.new do |*args, &k|
+      self.run(*args) do |*args2|
+        next_cps = pm.call(*args2)
+        next_cps.is_a?(Cps) or raise "`and_then` block did not return a Cps. Did you want `and_lift`? at #{pm.source_location}"
```

Comment:
I replaced the backticks with single quotes. We're not in markdown :)

Putting a message in a variable as is often done allocates a string every time. If you're going to put it in a variable, only do so when you actually need it (i.e., in the error case).

I'd argue that this is one of those cases where the end of the line doesn't matter in ordinary reading of the code. The relevant part is `next_cps.is_a?(Cps) or raise`. 99% of the time, I don't care what the message is. In such cases, I don't mind if the line is too long to read in a normal width window. I considered moving the `raise "..."` into a method, but that just spreads the code out. It's a private method, so do I put it at the bottom of the file? It doesn't make it any more readable. I'm leaving as is.